src/Future.ts

```diff
@@ -31,59 +19,89 @@ export class Future<T> {
-        return new Future(Lazy.of(() => new Promise<T[]>(
-            (resolve,reject) => cb((v:T) => resolve([v])))));
+    static ofPromiseCtor<T>(executor: (resolve:(x:T)=>void, reject: (x:any)=>void)=>void): Future<T> {
+        return new Future(new Promise(executor).then(v=>[v]));
     }
 
     /**
      * Build a Future from an existing javascript Promise.
      */
     static of<T>(promise: Promise<T>): Future<T> {
-        return new Future(Lazy.of(()=>promise.then(x => [x])));
+        return new Future(promise.then(x => [x]));
+    }
+
+    /**
+     * Build a Future from a node-style callback API, for instance:
+     *
+     *     Future.ofCallback<string>(cb => fs.readFile('/etc/passwd', 'utf-8', cb))
+     */
+    static ofCallback<T>(fn: (cb:(err:any, val:T)=>void)=>void): Future<T> {
+        return Future.ofPromiseCtor((resolve,reject)=>fn((err, data)=>{
+            if (err) {
+                reject(err);
+            } else {
+                resolve(data);
+            }
+        }));
     }
 
     /**
      * Build a successful Future with the value you provide.
      */
     static ok<T>(val:T): Future<T> {
-        return new Future(Lazy.of(()=>Promise.resolve([val])));
+        return new Future(Promise.resolve([val]));
     }
 
     /**
      * Build a failed Future with the error data you provide.
      */
     static failed<T>(reason: any): Future<T> {
-        return new Future(Lazy.of(()=>Promise.reject(reason)));
+        return new Future(Promise.reject(reason));
+    }
+
+    /**
+     * Creates a Future from a function returning a Promise,
+     * which can be inline in the call, for instance:
+     *
+     *     const f1 = Future.ok(1);
+     *     const f2 = Future.ok(2);
+     *     return Future.do(async () => {
+     *         const v1 = await f1;
+     *         const v2 = await f2;
+     *         return v1 + v2;
+     *     });
+     */
+    static do<T>(fn: ()=>Promise<T>): Future<T> {
+        return Future.of(fn())
     }
 
     /**
      * The `then` call is not meant to be a part of the `Future` API,
      * we need then so that `await` works directly.
-     * This method is eager, will trigger the underlying Promise.
      *
      * Please rather use [[Future.map]] or [[Future.flatMap]].
      */
     then<TResult1 = T, TResult2 = never>(
         onfulfilled: ((value: T) => TResult1 | PromiseLike<TResult1>),
         onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): PromiseLike<TResult1 | TResult2> {
-        return this.promise.get().then(([x]) => onfulfilled(x), rejected => onrejected?onrejected(rejected):Promise.reject(rejected)); 
+        return this.promise.then(([x]) => onfulfilled(x), rejected => onrejected?onrejected(rejected):Promise.reject(rejected)); 
     }
 
     /**
      * Get a `Promise` from this `Future`.
      */
     toPromise(): Promise<T> {
-        return this.promise.get().then(([x]) => x);
+        return this.promise.then(([x]) => x);
```

Comment:
Could we store a reference to the original promise and return that instead?
```ts
const unit = Promise.resolve()
const a = Future.of(Future.of(unit).toPromise()).toPromise()
const b = Future.do(()=>unit).toPromise()
console.log(...[a,b].map(x=>x===unit))
// > false false
```
😱 :scream:

Reply:
hmm I'm not sure if that's really important it's the same Promise. We don't get much from that, but if we must keep the original promise then we do waste some memory potentially. The fact it's a different promise is the reason that I named it `toPromise` and not `getPromise` (the apidoc also says 'get a Promise', not 'get the Promise'). And also the fact that there's an underlying promise is (theoritically at least) an implementation detail.

One thing that could trip a user could be more the fact that if you call like three times `toPromise` you get three different promises.

But I still don't know whether that's worth the extra memory use of keeping the original promise hanging around. I mean it's not much memory, it's just if you have a list of tens of thousands of futures or something.. I don't know, do you think it's worth the memory use?

Reply:
Not sure. It's (keeping 2 Promises around vs one) vs (creating a promise for every call to toPromise vs giving out references to the same one)

first harms
```ts
const f = Future.of(Promise.resolve(0))
new Array(100000).fill(0).map(_=>f.toPromise())
// 2 Promises
```
```ts
new Array(100000).fill(0).map(_=>Future.of(Promise.resolve(_))
// 200000 Promises
```
while the second harms
```ts
const f = Future.of(Promise.resolve(0))
new Array(100000).fill(0).map(_=>f.toPromise())
// 100001 Promises
```
```ts
new Array(100000).fill(0).map(_=>Future.of(Promise.resolve(_))
// 100000 Promises
```